doc/rfc/781-lua-system-wide-profiler.md

```diff
@@ -0,0 +1,395 @@
+# lua: system-wide profiler
+
+* **Status**: Done
+* **Start date**: 02-07-2021
+* **Authors**: Mikhail Shishatskiy @Shishqa m.shishatskiy@tarantool.org, Maxim Kokryashkin @fckxorg m.kokryashkin@tarantool.org
+* **Issues**: [#781](https://github.com/tarantool/tarantool/issues/781), [#4001](https://github.com/tarantool/tarantool/issues/4001)
+
+## Summary
+The document describes the platform profiler for LuaJIT. It is needed to obtain a complete view of platform performance. The existing LuaJIT profiler is only able to give you information about virtual machine states and guest stacks. Hence, the document proposes to provide a new sampling profiler, so it will be able to gather stack traces from both C and Lua.
+
+## Background and motivation
+
+Currently, the available options for profiling LuaJIT are not fine enough to provide an understanding of performance. For example, perf is only able to show the host stack, so all the Lua calls are seen as a single pcall. Oppositely, the jit.p module provided with LuaJIT is not able to give any information about the host stack.
+
+To get a detailed perspective on the platform's performance, a more advanced profiler is needed. The desired profiler must be able to capture both guest and host stacks simultaneously, along with virtual machine states.
+
+Assume we have the following chunks of code:
+```c
+1  // file c_payload.c
+2
+3  // fibonnacci number
+4  double c_payload(double n) {
+5    if (n <= 1) {
+6      return n;
+7    }
+8    return c_payload(n - 1) + c_payload(n - 2);
+9  }
+```
+
+```lua
+ 1  -- file lua_payload.lua
+ 2
+ 3  local jit = require("jit")
+ 4  local ffi = require("ffi")
+ 5  local os = require("os")
+ 6
+ 7  ------------------------------------------------
+ 8
+ 9  -- fibonacci number
+10  function lua_payload(n)
+11    if n <= 1 then
+12      return n
+13    end
+14    return lua_payload(n - 1) + lua_payload(n - 2)
+15  end
+16
+17  -- mind project-specific package location --
+18  local c = ffi.load("c_payload")
+19
+20  ffi.cdef[[
+21  double c_payload(double n);
+22  ]]
+23
+24  ------------------------------------------------
+25
+26  local start_time = os.clock()
+27  local benchmark_time = 10
+28
+29  jit.off()
+30  jit.flush()
+31
+32  local i = 0
+33  while os.clock() - start_time < benchmark_time do
+34      if i % 2 == 0 then
+35          c.c_payload(20)
+36      else
+37          lua_payload(20)
+38      end
+39      i = i + 1
+40  end
+```
+
+The `c_payload.c` code was compiled into a shared library and loaded by LuaJIT via the FFI.
+
+To see the difference, you can take a look at the flamegraphs (learn more about [flamegraphs](https://www.brendangregg.com/flamegraphs.html)) for this code generated by perf, jit.p, and PoC for the proposed profiler below.
+
+### perf
+![perf](https://i.imgur.com/DlKbFpo.png)
```

Comment:
Third party services may gone. I suggest to add a directory and place the images right here. You can use a relative URI: `![perf](images/perf.png)`.